src/index.js

```diff
@@ -170,224 +172,232 @@ class UglifyJsPlugin {
+                const map = asset.map(this.options);
 
-              input = source;
+                if (UglifyJsPlugin.isSourceMap(map)) {
+                  inputSourceMap = map;
+                } else {
+                  inputSourceMap = map;
 
-              if (UglifyJsPlugin.isSourceMap(map)) {
-                inputSourceMap = map;
+                  compilation.warnings.push(
+                    new Error(`${file} contains invalid source map`)
+                  );
+                }
               } else {
-                inputSourceMap = map;
-
-                compilation.warnings.push(
-                  new Error(`${file} contains invalid source map`)
-                );
+                inputSourceMap = null;
               }
-            } else {
-              input = asset.source();
-              inputSourceMap = null;
-            }
 
-            // Handling comment extraction
-            let commentsFile = false;
+              // Handling comment extraction
+              let commentsFile = false;
 
-            if (this.options.extractComments) {
-              commentsFile =
-                this.options.extractComments.filename || `${file}.LICENSE`;
+              if (this.options.extractComments) {
+                commentsFile =
+                  this.options.extractComments.filename || `${file}.LICENSE`;
 
-              if (typeof commentsFile === 'function') {
-                commentsFile = commentsFile(file);
+                if (typeof commentsFile === 'function') {
+                  commentsFile = commentsFile(file);
+                }
               }
-            }
 
-            const task = {
-              file,
-              input,
-              inputSourceMap,
-              commentsFile,
-              extractComments: this.options.extractComments,
-              uglifyOptions: this.options.uglifyOptions,
-              minify: this.options.minify,
-            };
-
-            if (this.options.cache) {
-              const defaultCacheKeys = {
-                // eslint-disable-next-line global-require
-                'uglify-js': uglifyJsPackageJson.version,
-                // eslint-disable-next-line global-require
-                'uglifyjs-webpack-plugin': require('../package.json').version,
-                'uglifyjs-webpack-plugin-options': this.options,
-                hash: crypto
-                  .createHash('md4')
-                  .update(input)
-                  .digest('hex'),
+              const task = {
+                file,
+                input,
+                inputSourceMap,
+                commentsFile,
+                extractComments: this.options.extractComments,
+                uglifyOptions: this.options.uglifyOptions,
+                minify: this.options.minify,
+                cacheKey,
               };
 
-              task.cacheKeys = this.options.cacheKeys(defaultCacheKeys, file);
```

Comment:
I moved the computation of the cache keep up so that we can check whether the key exists in the cache before computing the input source map.